### PR TITLE
PUBDEV-7138 - Extended Isolation Forest

### PR DESCRIPTION
**Pull request by valenad1**
Implementation of Extended Isolation Forest algorithm.

https://0xdata.atlassian.net/browse/PUBDEV-7138

Suggestions where to look:
* ExtendedIsolationForestTest
* ExtendedIsolatonForest -> MAX_NTREES, MAX_SAMPLE_SIZE
* SamplingUtils
* ArrayUtils.gaussianVector()
* h2o-py/tests/testdir_algos/isoforextended/
* h2o-r/tests/testdir_algos/isoforextended

### Algorithm input
Only attributes listed here [h2o-docs/src/product/data-science/eif.rst](https://github.com/h2oai/h2o-3/blob/valenad-PUBDEV-7138-extended-isolation-forest/h2o-docs/src/product/data-science/eif.rst) are supported.
```
eif_model = H2OExtendedIsolationForestEstimator(ntrees=100,
                                                sample_size=256,
                                                extension_level=0,
                                                categorical_encoding="enum",
                                                seed=0xBEEF)
eif_model.train(training_frame=train)
```

### Whole algorithm output from training

```
Model Details
=============
H2OExtendedIsolationForestEstimator :  Extended Isolation Forest
Model Key:  ExtendedIsolationForest_model_python_1616456143566_3


Model Summary: 
    number_of_trees    size_of_subsample    extension_level    seed
--  -----------------  -------------------  -----------------  ------
    100                256                  0                  48879

```

### Algorithm output from scoring

```
  anomaly_score    mean_length
---------------  -------------
       0.744915        4.34919
       0.367192       14.7964
```

## Computing time benchmark compared to H2O Isolation Forest on Creditcards dataset - training

h2o-train-perf_credit0 37829865057254497

## Computing time benchmark compared to H2O Isolation Forest on Creditcards dataset - scoring

h2o-score-perf_credit0 4507518975762984

### Reported on validation dataset

all-valid-0 217096651659068

### Reported on testing dataset

all-0 6544134222570593

### CODE REVIEW

File: h2o-algos/src/test/java/hex/tree/isoforextended/ExtendedIsolationForestTest.java
Comment (maurever) [inline h2o-algos/src/test/java/hex/tree/isoforextended/ExtendedIsolationForestTest.java:390]: Unused code.

File: h2o-core/src/main/java/water/util/FrameUtils.java
Comment (maurever) [inline h2o-core/src/main/java/water/util/FrameUtils.java:183]: Maybe the number 100 000 could be some constant defined globally?

**Review by maurever — COMMENTED** [review]
LGTM, just a few suggestions.

File: h2o-algos/src/main/java/hex/tree/isoforextended/isolationtree/CompressedIsolationTree.java
Comment (valenad1) [reply h2o-algos/src/main/java/hex/tree/isoforextended/isolationtree/CompressedIsolationTree.java:37]: It shouldn't be pain for code or computing time, I did it.

File: h2o-algos/src/test/java/hex/tree/isoforextended/ExtendedIsolationForestTest.java
Comment (valenad1) [reply h2o-algos/src/test/java/hex/tree/isoforextended/ExtendedIsolationForestTest.java:390]: I replaced it with comment

File: h2o-core/src/main/java/water/util/FrameUtils.java
Comment (valenad1) [reply h2o-core/src/main/java/water/util/FrameUtils.java:183]: I created FrameUtils.MAX_VEC_NUM_ROWS_FOR_ARRAY_EXPORT constant for this purpose

**Review by maurever — APPROVED** [review]

**Review by michalkurka — APPROVED** [review]
Looks great, thank you @valenad1

File: h2o-algos/src/main/java/hex/tree/isoforextended/SamplingUtils.java
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/tree/isoforextended/SamplingUtils.java:41]: typically we try to avoid serializables in MRTasks

File: h2o-algos/src/main/java/hex/tree/isoforextended/SamplingUtils.java
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/tree/isoforextended/SamplingUtils.java:22]: I would rather remove this if it is not used

File: h2o-core/src/main/java/water/util/ArrayUtils.java
Comment (michalkurka) [inline h2o-core/src/main/java/water/util/ArrayUtils.java:1154]: it is not usual for H2O to work with Sets, instead arrays and primite objects are typically used

File: h2o-core/src/main/java/water/util/ArrayUtils.java
Comment (michalkurka) [inline h2o-core/src/main/java/water/util/ArrayUtils.java:1140]: imho this function shouldn't be here - it is too specific to your use case

instead of putting all the functionality into one function, keep the original gaussianVector

add a function that randomly zero-s out a portion of a row
and call them in EIF

File: h2o-core/src/main/java/water/util/ArrayUtils.java
Comment (michalkurka) [inline h2o-core/src/main/java/water/util/ArrayUtils.java:936]: doesn't work for empty array
doesn't work for NAs

File: h2o-core/src/main/java/water/util/ArrayUtils.java
Comment (michalkurka) [inline h2o-core/src/main/java/water/util/ArrayUtils.java:304]: I don't understand this method (including method name)

**Comment by michalkurka** [comment]
Please resolve the comments before merging but approved from my side assuming the changes will be made

**Comment by michalkurka** [comment]
One thing to try to add (in another PR perhaps) - grid search test

File: h2o-core/src/main/java/water/util/ArrayUtils.java
Comment (valenad1) [reply h2o-core/src/main/java/water/util/ArrayUtils.java:304]: Never mind, I can replace it with standard add method.

**Comment by valenad1** [comment]
🚀

**Pull request merged**
